src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.EventDispatch.cs

```diff
@@ -78,9 +78,16 @@ private void SetNotFoundResponse(object? sender, EventArgs args)
     {
         if (_httpContext.Response.HasStarted)
         {
-            throw new InvalidOperationException("Cannot set a NotFound response after the response has already started.");
+            // We cannot set a NotFound code after the response has already started
+            var navigationManager = _httpContext.RequestServices.GetRequiredService<NavigationManager>();
+            var notFoundUri = $"{navigationManager.BaseUri}not-found";
+            navigationManager.NavigateTo(notFoundUri);
```

Comment:
I would have expected we did this like what we do for redirections. https://github.com/dotnet/aspnetcore/blob/8ddf53d847a8ce1cb0e0787f465ad1aa3ce3ecf1/src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Streaming.cs#L227-L228

Send the not found event to the browser and let the JS side deal with it.

Reply:
The reason for this is that we don't want to change the URL, we just want to display the not found contents. The JS side can essentially issue a `fetch` request to the server to render the page and overwrite the existing contents, which is the equivalent of what happens when we call `NotFound()` in interactive server/webassembly.